cmd/snap-seccomp/main.go

```diff
@@ -619,21 +623,39 @@ func compile(content []byte, out string) error {
 	for scanner.Scan() {
 		line := strings.TrimSpace(scanner.Text())
 
-		// FIXME: right now complain mode is the equivalent to
-		// unrestricted.  We'll want to change this once we
-		// seccomp logging is in order.
-		//
 		// special case: unrestricted means we switch to an allow-all
 		//               filter and are done
-		if line == "@unrestricted" || line == "@complain" {
+		if line == "@unrestricted" {
 			secFilter, err = seccomp.NewFilter(seccomp.ActAllow)
 			if err != nil {
-				return fmt.Errorf("cannot create seccomp filter: %s", err)
+				return fmt.Errorf("cannot create unrestricted seccomp filter: %s", err)
 			}
 			if err := addSecondaryArches(secFilter); err != nil {
 				return err
 			}
 			break
+		} else if line == "@complain" {
+			var fallback bool = false
+
+			secFilter, err = seccomp.NewFilter(seccomp.ActLog)
```

Comment:
While it would be a bug in the policy to have '@complain' in it more than once, the previous code stopped processing rules as soon as it found '@complain', but this code will add a NewFilter for every '@complain'. I think we should simply skip (perhaps logging with debug) additional '@complain' lines.

Reply:
It is no different than if one or more white listed system calls are placed above the first `@complain` directive. In that case, a new filter would be allocated when `@complain` was encountered and all the previously white listed system calls would be forgotten.

If we're going to handle multiple `@complain` directives, we should handle the case that I described, too. Unfortunately, libseccomp doesn't allow us to change the default action of an existing filter via `seccomp_attr_set(..., SCMP_FLTATR_ACT_DEFAULT, SCMP_ACT_LOG)` so an entirely new filter must be allocated.

What is your preferred solution here?
1. Don't worry about the case that I mentioned. Address your concern by skipping `NewFilter()` when libseccomp-golang's `GetDefaultAction()` returns `ActLog`.
1. Treat both situations as an error. (I don't know what it means to error out of the `compile()` function.)

Reply:
What snap-confine would do before the move to snap-seccomp is it would preprocess the file, looking for @unrestricted or @complain. The rewrite took a shortcut and simply says "forget anything I did before if I find either of these". That was fine before when they were equivalent. Profile authors will sometimes put these directives anywhere in the file, so I think we need to account for that.

Instead of the two listed options, I suggest the previous behavior: preprocess the file, if @unrestricted found 1 or more times, store that off, if @complain found 1 or more times, store that off. Eg something along the lines of:
```
// Preprocess to allow @unrestricted and @complain to appear anywhere in the profile
unrestricted, complain := preprocess()
if unrestricted {
    seccomp.NewFilter(seccomp.ActAllow)
} else if complain {
    seccomp.NewFilter(seccomp.ActLog)
} else {
    seccomp.NewFilter(seccomp.ActErrno.SetReturnCode(C.EPERM))
}

addSecondaryArches(secFilter)
if not unrestricted and not complain {
    for scanner.Scan() {
        line := strings.TrimSpace(scanner.Text())
        parseLine(line, secFilter)
        ...
    }
}
...
```

This has the added benefit of adding the secondary arch in one spot.

@mvo5 - what do you think about preprocessing?